src/osd/OSD.cc

```diff
@@ -1455,6 +1459,25 @@ void OSDService::queue_for_peering(PG *pg)
   peering_wq.queue(pg);
 }
 
+void OSDService::check_completion_item() {
+  while (1) {
+    // dout(0) << __func__ << " " << __LINE__ << " enter!! " << dendl;
+    if (!osd) assert(0);
+
+    if (comp_thread_stop) {
+      dout(0) << __func__ << " " << __LINE__ << " out " << dendl;
+      break;
+    }
+    RWLock::RLocker l(osd->pg_map_lock);
+    ceph::unordered_map<spg_t, PG*>::iterator iter = osd->pg_map.begin();
```

Comment:
@athanatos Need to iteration but not once/sec. this polling function can be changed to event-driven approach. (sleep until happening events). Note that another completion function for processing completion is in execute_ctx. Therefore, most of processing completion will be handled at execute_ctx which already hold pg lock (at issuing context).  This is for preventing unfinished jobs.